docs: add Airbyte Source Connector FAQ (#7139)
meltano/meltano: docs/src/_guide/advanced-topics.md

* `s3` - Support for using S3 as a [state backend](/concepts/state_backends)
* `gcs` - Support for using Google Cloud Storage as a [state backend](/concepts/state_backends)
* `azure` - Support for using Azure Blob Storage as a [state backend](/concepts/state_backends)

## Airbyte Connector Integration FAQ

### How do the Singer and Airbyte specifications relate?

> The Singer specification was started in 2016 by Stitch Data. It specified a data transfer format that would allow any number of data systems, called taps, to send data to any data destination, called targets. Airbyte was incorporated in 2020 and created their own specification that was heavily inspired by Singer. There are differences but the core of each specification is sending new-line delimited JSON data from STDOUT of a tap to STDIN of a target.

Reviewer: ```suggestion
The Singer specification was started in 2016 by Stitch Data. It specified a data transfer format that would allow any number of data systems, called taps, to send data to any data destinations, called targets. Airbyte was incorporated in 2020 and created their own specification that was heavily inspired by Singer. There are differences, but the core of each specification is sending new-line delimited JSON data from STDOUT of a tap to STDIN of a target.
```

### How does the Airbyte Connector with Meltano integration work?

A community member used the Meltano Singer SDK to write tap-airbyte-wrapper. This wrapper connector calls the Docker image for a given Airbyte Source and translates the messages into a Singer-compatible format. The output of the tap is conformed to the Singer standard and then can be sent to any Singer target, many of which are listed on [MeltanoHub](hub.meltano.com/loaders/).

### How do I get help if I need it?

We first recommend you read through this FAQ to see if your question has been answered. We next recommend searching the [Meltano docs](https://docs.meltano.com) to see if your answer is there.

If you still need help we’d recommend either filing an issue in the [Meltano Github repository](https://github.com/meltano/meltano/issues) or join our [Slack community](https://meltano.com/slack) to ask there.

We also have [weekly Office Hours](https://www.addevent.com/calendar/Li390615) where you can talk to the Meltano team and ask any questions you have to us!

> ### Does this wrap the Airbyte UI or API?

Reviewer 2: ```suggestion
### Do I need to have an Airbyte UI or API instance to use this?
```

I think the main point this should address is whether someone needs to run Airbyte and Meltano to use this. We want them to know that Meltano is directly running the connectors without the Airbyte interface.

> No. This integration only makes it possible to run Airbyte [Source Connectors](https://docs.airbyte.com/category/sources) within your Meltano project.

Reviewer 2: ```suggestion
No. This integration makes it possible to directly run Airbyte [Source Connectors](https://docs.airbyte.com/category/sources) within your Meltano project. There's no need to run the Airbyte UI or API to use this feature.
```

### Does this support Airbyte destination connectors as well?

Airbyte destinations are _not_ supported with tap-airbyte-wrapper. This is a feature that could be added in the future.

### What is the recommended way to install and use these connectors?

> Any connector that is listed on MeltanoHub as being maintained by `airbyte` can be installed and run as you would any other connector. Use `meltano add extractor &lt;tap>` to add and install.

Reviewer 2: ```suggestion
Any connector that is listed on MeltanoHub as being maintained by `airbyte` can be installed and run as you would any other connector. Use `meltano add extractor <tap>` to add and install.
```

It looks like the angle bracket doesnt need to be encoded based on the netlify preview. I havent tested this though.

### What are the limitations of running Airbyte-based connectors with Meltano?

Based on the current implementation and testing, the main challenge is putting these connectors into production with Meltano. See the next question for more details on this.

### Am I able to put my Meltano project with Airbyte connectors in production?

> Yes, assuming you can set up Docker-in-docker. Since each Airbyte connector is itself a Docker image, running inside a dockerized Meltano project would require “docker in docker” which has some possible challenges around privileges and permissions on certain systems.

Reviewer 2: This assumes that production deployments require Docker which isnt necessarily true. We might want to consider prefacing that the challenges only come with using Dockerized meltano projects. If someone doenst package with a docker image and the production machine has docker available then this just works.

Author: Clarified. Feel free to make direct edits if you feel there is a better way to say it.

Within Codespaces, adding support for docker-in-docker was as easy as [adding a few lines](https://github.com/meltano/meltano-codespace-ready/commit/5ffb9fb6e3232142c8e2307340c0a4fb66379db4) to the devcontainer.json file.

It is also possible to run [Meltano on GitHub Actions](https://github.com/brooklyn-data/meltano-on-github-actions). It’s likely possible to update this action to supporting docker-in-docker as well.

> There [are](zhsj.me/blog/view/dind-without-privileged) [several](shisho.dev/blog/posts/docker-in-docker) [articles](docs.docker.com/engine/security/rootless) on the web which discuss docker-in-docker in more detail.

Reviewer: It may not be obvious that these words link to different sites

> We are interested in exploring this challenge with the community. Please share in **this discussion or connect with us in Slack**about your experiences with attempting to put this into production.

Reviewer: ```suggestion
We are interested in exploring this challenge with the community. Please share in **this discussion or connect with us in Slack** about your experiences with attempting to put this into production.
```

### How does this work with custom connectors made with the Airbyte CDK?

This integration will work with any dockerized Airbyte source, whether it was made with their CDK or not.

To configure your Airbyte connector as a [custom plugin]([https://docs.meltano.com/concepts/project#custom-plugin-definitions](https://docs.meltano.com/concepts/project#custom-plugin-definitions)) in Meltano you can copy the configuration shown below into your meltano.yml file then replace `name`, and the `value` for the airbyte_spec.image to reference your docker image. You can configure your connector without defining all of the settings and their metadata.

```yaml
 - name: tap-pokeapi # REPLACE THIS WITH YOUR CONNECTOR NAME
   variant: airbyte
   executable: tap-airbyte
   namespace: tap_airbyte
   pip_url: git+https://github.com/z3z1ma/tap-airbyte.git
   capabilities:
   - catalog
   - state
   - discover
   - about
   - stream-maps
   - schema-flattening
   settings:
   - description: Airbyte image to run
     kind: string
     label: Airbyte Spec Image
     name: airbyte_spec.image

>      value: airbyte/source-punk-api # REPLACE THIS WITH YOUR IMAGE NAME

Reviewer 2: ```suggestion
     value: airbyte/source-pokeapi # REPLACE THIS WITH YOUR IMAGE NAME
```

This was my fault but its probably less confusing if the example tap name and docker image match.

   - description: Airbyte image tag
     kind: string
     label: Airbyte Spec Tag
     name: airbyte_spec.tag
     value: latest
   - [INSERT OTHER SETTINGS HERE]
```

### What features does this add to Airbyte connectors?

Running Airbyte sources with Meltano brings a number of benefits to those connectors. With Meltano you get the ability to run stream maps to adjust data on the fly, you can define multiple environments to override configuration depending on where the EL pipeline is run, and you get the benefits of version control since everything is defined in your meltano.yml file.

We also recently added support for alternative [state backends](https://docs.meltano.com/concepts/state_backends) to Meltano. This means if you want to store the incremental state between runs in a place other than your system database or local filesystem, you can! Currently Meltano supports AWS S3, Azure Blob Storage, and Google Cloud Storage as alternative state backends.

> Since tap-airbyte-wrapper was written with the SDK, this also unlocks the `BATCH` message format which can help with overall pipeline throughput to compatible targets.

Reviewer: Has `BATCH` been tested with this? I'd be elated if it works out-of-the-box 😅

Author: @edgarrmondragon it does! It's not guaranteed to necessarily be faster since the tap still serializes it, but it does work!

There are tradeoffs for both of these methods and you would have to determine where to best invest your time depending on your needs.

If you have other ways to make this transition easier, we’d love to discuss them in Slack or Office Hours with you!

### Do these connectors work with `meltano run` and `meltano elt`?

Yes! Since the wrapper is based on the Meltano Singer SDK these connectors work just as they would a python-based Singer connector. You can [run](https://docs.meltano.com/reference/command-line-interface#run), [elt](https://docs.meltano.com/reference/command-line-interface#elt), [invoke](https://docs.meltano.com/reference/command-line-interface#invoke), [test](https://docs.meltano.com/reference/command-line-interface#test), and [configure](https://docs.meltano.com/reference/command-line-interface#config) them as you normally would.

### Does this integration support stream maps for PII filtering/hashing?

> Yes! This is a unique feature of Meltano that is not found in as comprehensive manner in other modern EL tools. With stream maps users are able to hash, filter, duplicate, and alias any column or table during their extract. Because this Airbyte integration was built using the SDK, streams maps can be defined on the connector itself ([SDK docs](https://sdk.meltano.com/en/latest/stream_maps.html)), or they can be added separately as a mapper in between a tap and target executing using `meltano run`. See [our tutorial](https://docs.meltano.com/getting-started/part4#undefined) for an example.

Reviewer 3: > found in as comprehensive manner in other

found in a comprehensive manner in other

Author: Fixed

* [Run the Docker daemon as a non-root user](https://docs.docker.com/engine/security/rootless/)

We are interested in exploring this challenge with the community. Please share in [this discussion](https://github.com/meltano/meltano/discussions/7142) or connect with us in [Slack](https://meltano.com/slack) about your experiences with attempting to put this into production.

### How does this work with custom connectors made with the Airbyte CDK?

This integration will work with any dockerized Airbyte source, whether it was made with their CDK or not.

To configure your Airbyte connector as a [custom plugin]([https://docs.meltano.com/concepts/project#custom-plugin-definitions](https://docs.meltano.com/concepts/project#custom-plugin-definitions)) in Meltano you can copy the configuration shown below into your meltano.yml file then replace `name`, and the `value` for the airbyte_spec.image to reference your docker image. You can configure your connector without defining all of the settings and their metadata.

```yaml
 - name: tap-pokeapi # REPLACE THIS WITH YOUR CONNECTOR NAME
   variant: airbyte
   executable: tap-airbyte
   namespace: tap_airbyte

>    pip_url: git+https://github.com/z3z1ma/tap-airbyte.git

Reviewer 2: ```suggestion
   pip_url: git+https://github.com/MeltanoLabs/tap-airbyte.git
```
